### PR TITLE
adding password option to ssh, scp and sftp: continued

### PR DESCRIPTION
**Pull request by mrk-its**
It seems that author abandoned original #292 PR, so I'll continue it here. I've addressed comments, @mpenkov please take a look.

Additionally I'm calling urlparse.unquote on username and password, allowing percent encoding of username / password.

### CODE REVIEW

File: smart_open/smart_open_lib.py [hunk 1]
```diff
@@ -554,6 +554,8 @@ def _open_binary_stream(uri, mode, transport_params):
                 host=parsed_uri.host,
                 user=parsed_uri.user,
                 port=parsed_uri.port,
+                password=parsed_uri.password,
+                transport_params=transport_params
```
Comment (mpenkov) [inline smart_open/smart_open_lib.py]:
```suggestion
                transport_params=transport_params,
```

File: setup.py [hunk 2]
```diff
@@ -51,7 +51,8 @@ def read(fname):
     # 'botocore < 1.11.0'
     # Not used directly but allows boto GCE plugins to load.
     # https://github.com/GoogleCloudPlatform/compute-image-packages/issues/262
-    'google-compute-engine==2.8.12'
+    'google-compute-engine==2.8.12',
+    'paramiko==2.6.0',
```
Comment (mpenkov) [inline setup.py]: Why is the version pin necessary?

Comment (mpenkov) [reply setup.py]: If we include paramiko in setup.py, then this part becomes irrelevant:

https://github.com/RaRe-Technologies/smart_open/blob/1770807b0c8da2a34f749afa17ff70a715bc85de/smart_open/ssh.py#L43-L50

Could you please clean it up?

File: smart_open/ssh.py [hunk 3]
```diff
@@ -87,11 +94,16 @@ def open(path, mode='r', host=None, user=None, port=DEFAULT_PORT):
     If you specify a previously unseen host, then its host key will be added to
     the local ~/.ssh/known_hosts *automatically*.
 
+    If `username` or `password` are specified in _both_ the uri and `transport_params`,
```
Comment (mpenkov) [inline smart_open/ssh.py]: This is ReST, you need double backticks, e.g. 

```
``username``
```

For emphasis, use

```
*emphasis*
```

File: smart_open/tests/test_smart_open.py [hunk 4]
```diff
@@ -231,14 +249,17 @@ class SmartOpenHttpTest(unittest.TestCase):
     @mock.patch('smart_open.ssh.open')
     def test_read_ssh(self, mock_open):
         """Is SSH line iterator called correctly?"""
-        obj = smart_open.smart_open("ssh://ubuntu@ip_address:1022/some/path/lines.txt")
+        obj = smart_open.smart_open("ssh://ubuntu:pass@ip_address:1022/some/path/lines.txt",
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]: Please use hanging indent here and everywhere. Alternatively, shorten the dummy data somehow.

File: smart_open/tests/test_smart_open.py [hunk 5]
```diff
@@ -231,14 +249,17 @@ class SmartOpenHttpTest(unittest.TestCase):
...
         obj.__iter__()
         mock_open.assert_called_with(
             '/some/path/lines.txt',
             'rb',
             host='ip_address',
             user='ubuntu',
-            port=1022
+            password='pass',
+            port=1022,
+            transport_params={'hello': 'world'}
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]:
```suggestion
            transport_params={'hello': 'world'},
```

File: smart_open/tests/test_ssh.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import mock
+import unittest
+
+import smart_open.ssh
+
+
+def mock_ssh(func):
+    def wrapper(*args, **kwargs):
+        smart_open.ssh._SSH.clear()
+        return func(*args, **kwargs)
```
Comment (mpenkov) [inline smart_open/tests/test_ssh.py]: A newline would help readability here.

```suggestion
        return func(*args, **kwargs)

```

File: smart_open/tests/test_smart_open.py [hunk 7]
```diff
@@ -196,22 +196,24 @@ def test_leading_slash_local_file(self):
         self.assertEqual(uri.uri_path, '//' + path)
 
     def test_ssh(self):
-        as_string = 'ssh://user@host:1234/path/to/file'
+        as_string = 'ssh://user:pass@host:1234/path/to/file'
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]: Please add an additional test for when the password is _not_ specified.

File: smart_open/tests/test_smart_open.py [hunk 8]
```diff
@@ -196,22 +196,24 @@ def test_leading_slash_local_file(self):
...
         self.assertEqual(uri.scheme, 'ssh')
         self.assertEqual(uri.uri_path, '/path/to/file')
         self.assertEqual(uri.user, 'user')
         self.assertEqual(uri.host, 'host')
         self.assertEqual(uri.port, 1234)
+        self.assertEqual(uri.password, 'pass')
 
     def test_scp(self):
-        as_string = 'scp://user@host:/path/to/file'
+        as_string = 'scp://user:pass@host:/path/to/file'
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]: Please add an additional test for when the password is not specified.

File: setup.py (hunk 2 above)
Comment (mrk-its) [reply setup.py]: paramiko is included only in test dependencies (and this is probably ok, we do not want to add such heavy package to main deps)

Comment (mrk-its) [reply setup.py]: > Why is the version pin necessary?

It isn't, removed.

**Comment by mpenkov** [comment]
Merged. Great work @mrk-its . Congrats on your first contribution to smart_open :1st_place_medal: !!

**Pull request merged**